File: RxTools-library/src/main/java/com/vondear/rxtools/view/progressing/animation/SpriteAnimatorBuilder.java
package com.vondear.rxtools.view.progressing.animation;

import android.animation.Keyframe;
import android.animation.ObjectAnimator;
import android.animation.PropertyValuesHolder;
import android.util.Property;
import android.view.animation.Animation;
import android.view.animation.Interpolator;

import com.vondear.rxtools.view.progressing.animation.interpolator.KeyFrameInterpolator;
import com.vondear.rxtools.view.progressing.sprite.Sprite;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * @author vondear
 */
public class SpriteAnimatorBuilder {
    private Sprite sprite;
    private List<PropertyValuesHolder> propertyValuesHolders = new ArrayList<>();
    private Interpolator interpolator;
    private int repeatCount = Animation.INFINITE;
    private long duration = 2000;

    public SpriteAnimatorBuilder(Sprite sprite) {
        this.sprite = sprite;
    }

    public SpriteAnimatorBuilder scale(float fractions[], float... scale) {
        holder(fractions, Sprite.SCALE, scale);
        return this;
    }

    public SpriteAnimatorBuilder alpha(float fractions[], int... alpha) {
        holder(fractions, Sprite.ALPHA, alpha);
        return this;
    }

    @SuppressWarnings("unused")
    public SpriteAnimatorBuilder scaleX(float fractions[], float... scaleX) {
        holder(fractions, Sprite.SCALE, scaleX);
        return this;
    }

    public SpriteAnimatorBuilder scaleY(float fractions[], float... scaleY) {
        holder(fractions, Sprite.SCALE_Y, scaleY);
        return this;
    }

    public SpriteAnimatorBuilder rotateX(float fractions[], int... rotateX) {
        holder(fractions, Sprite.ROTATE_X, rotateX);
        return this;
    }

    public SpriteAnimatorBuilder rotateY(float fractions[], int... rotateY) {
        holder(fractions, Sprite.ROTATE_Y, rotateY);
        return this;
    }

    @SuppressWarnings("unused")
    public SpriteAnimatorBuilder translateX(float fractions[], int... translateX) {
        holder(fractions, Sprite.TRANSLATE_X, translateX);
        return this;
    }


    @SuppressWarnings("unused")
    public SpriteAnimatorBuilder translateY(float fractions[], int... translateY) {
        holder(fractions, Sprite.TRANSLATE_Y, translateY);
        return this;
    }


    public SpriteAnimatorBuilder rotate(float fractions[], int... rotate) {
        holder(fractions, Sprite.ROTATE, rotate);
        return this;
    }

    public SpriteAnimatorBuilder translateXPercentage(float fractions[], float... translateXPercentage) {
        holder(fractions, Sprite.TRANSLATE_X_PERCENTAGE, translateXPercentage);
        return this;
    }

    public SpriteAnimatorBuilder translateYPercentage(float[] fractions, float... translateYPercentage) {
        holder(fractions, Sprite.TRANSLATE_Y_PERCENTAGE, translateYPercentage);
        return this;
    }

    private PropertyValuesHolder holder(float[] fractions, Property property, float[] values) {
        ensurePair(fractions.length, values.length);
        Keyframe[] keyframes = new Keyframe[fractions.length];
        for (int i = 0; i < values.length; i++) {
            keyframes[i] = Keyframe.ofFloat(fractions[i], values[i]);
        }
        PropertyValuesHolder valuesHolder = PropertyValuesHolder.
                ofKeyframe(property
                        , keyframes
                );
        propertyValuesHolders.add(valuesHolder);
        return valuesHolder;
    }

    private PropertyValuesHolder holder(float[] fractions, Property property, int[] values) {
        ensurePair(fractions.length, values.length);
        Keyframe[] keyframes = new Keyframe[fractions.length];
        for (int i = 0; i < values.length; i++) {
            keyframes[i] = Keyframe.ofInt(fractions[i], values[i]);
        }
        PropertyValuesHolder valuesHolder = PropertyValuesHolder.
                ofKeyframe(property
                        , keyframes
                );
        propertyValuesHolders.add(valuesHolder);
        return valuesHolder;
    }

    private void ensurePair(int fractionsLength, int valuesLength) {
        if (fractionsLength != valuesLength) {
            throw new IllegalStateException(String.format(
                    Locale.getDefault(),
                    "The fractions.length must equal values.length, " +
                            "fraction.length[%d], values.length[%d]",
                    fractionsLength,
                    valuesLength));
        }
    }


    public SpriteAnimatorBuilder interpolator(Interpolator interpolator) {
        this.interpolator = interpolator;
        return this;
    }

    public SpriteAnimatorBuilder easeInOut(float... fractions) {
        interpolator(KeyFrameInterpolator.easeInOut(
                fractions
        ));
        return this;
    }


    public SpriteAnimatorBuilder duration(long duration) {
        this.duration = duration;
        return this;
    }

    @SuppressWarnings("unused")
    public SpriteAnimatorBuilder repeatCount(int repeatCount) {
        this.repeatCount = repeatCount;
        return this;
    }


    public ObjectAnimator build() {
        PropertyValuesHolder[] holders = new PropertyValuesHolder[propertyValuesHolders.size()];
        ObjectAnimator animator = ObjectAnimator.ofPropertyValuesHolder(sprite,
                propertyValuesHolders.toArray(holders));
        animator.setDuration(duration);
        animator.setRepeatCount(repeatCount);
        animator.setInterpolator(interpolator);
        return animator;
    }

}

File: RxUI/src/main/java/com/tamsiree/rxui/view/progressing/animation/SpriteAnimatorBuilder.kt
package com.tamsiree.rxui.view.progressing.animation

import android.animation.Keyframe
import android.animation.ObjectAnimator
import android.animation.PropertyValuesHolder
import android.util.Property
import android.view.animation.Animation
import android.view.animation.Interpolator
import com.tamsiree.rxui.view.progressing.animation.interpolator.KeyFrameInterpolator
import com.tamsiree.rxui.view.progressing.sprite.Sprite
import java.util.*

/**
 * @author tamsiree
 */
class SpriteAnimatorBuilder(private val sprite: Sprite) {
    private val propertyValuesHolders: MutableList<PropertyValuesHolder> = ArrayList()
    private var interpolator: Interpolator? = null
    private var repeatCount = Animation.INFINITE
    private var duration: Long = 2000
    fun scale(fractions: FloatArray, vararg scale: Float): SpriteAnimatorBuilder {
        holder(fractions, Sprite.SCALE, scale)
        return this
    }

    fun alpha(fractions: FloatArray, vararg alpha: Int): SpriteAnimatorBuilder {
        holder(fractions, Sprite.ALPHA, alpha)
        return this
    }

    fun scaleX(fractions: FloatArray, vararg scaleX: Float): SpriteAnimatorBuilder {
        holder(fractions, Sprite.SCALE, scaleX)
        return this
    }

    fun scaleY(fractions: FloatArray, vararg scaleY: Float): SpriteAnimatorBuilder {
        holder(fractions, Sprite.SCALE_Y, scaleY)
        return this
    }

    fun rotateX(fractions: FloatArray, vararg rotateX: Int): SpriteAnimatorBuilder {
        holder(fractions, Sprite.ROTATE_X, rotateX)
        return this
    }

    fun rotateY(fractions: FloatArray, vararg rotateY: Int): SpriteAnimatorBuilder {
        holder(fractions, Sprite.ROTATE_Y, rotateY)
        return this
    }

    fun translateX(fractions: FloatArray, vararg translateX: Int): SpriteAnimatorBuilder {
        holder(fractions, Sprite.TRANSLATE_X, translateX)
        return this
    }

    fun translateY(fractions: FloatArray, vararg translateY: Int): SpriteAnimatorBuilder {
        holder(fractions, Sprite.TRANSLATE_Y, translateY)
        return this
    }

    fun rotate(fractions: FloatArray, vararg rotate: Int): SpriteAnimatorBuilder {
        holder(fractions, Sprite.ROTATE, rotate)
        return this
    }

    fun translateXPercentage(fractions: FloatArray, vararg translateXPercentage: Float): SpriteAnimatorBuilder {
        holder(fractions, Sprite.TRANSLATE_X_PERCENTAGE, translateXPercentage)
        return this
    }

    fun translateYPercentage(fractions: FloatArray, vararg translateYPercentage: Float): SpriteAnimatorBuilder {
        holder(fractions, Sprite.TRANSLATE_Y_PERCENTAGE, translateYPercentage)
        return this
    }

    private fun holder(fractions: FloatArray, property: Property<*, *>, values: FloatArray): PropertyValuesHolder {
        ensurePair(fractions.size, values.size)
        val keyframes = arrayOfNulls<Keyframe>(fractions.size)
        for (i in values.indices) {
            keyframes[i] = Keyframe.ofFloat(fractions[i], values[i])
        }
        val valuesHolder = PropertyValuesHolder.ofKeyframe(property
                , *keyframes
        )
        propertyValuesHolders.add(valuesHolder)
        return valuesHolder
    }

    private fun holder(fractions: FloatArray, property: Property<*, *>, values: IntArray): PropertyValuesHolder {
        ensurePair(fractions.size, values.size)
        val keyframes = arrayOfNulls<Keyframe>(fractions.size)
        for (i in values.indices) {
            keyframes[i] = Keyframe.ofInt(fractions[i], values[i])
        }
        val valuesHolder = PropertyValuesHolder.ofKeyframe(property
                , *keyframes
        )
        propertyValuesHolders.add(valuesHolder)
        return valuesHolder
    }

    private fun ensurePair(fractionsLength: Int, valuesLength: Int) {
        check(fractionsLength == valuesLength) {
            String.format(
                    Locale.getDefault(),
                    "The fractions.length must equal values.length, " + "fraction.length[%d], values.length[%d]",
                    fractionsLength,
                    valuesLength)
        }
    }

    fun interpolator(interpolator: Interpolator?): SpriteAnimatorBuilder {
        this.interpolator = interpolator
        return this
    }

    fun easeInOut(vararg fractions: Float): SpriteAnimatorBuilder {
        interpolator(KeyFrameInterpolator.easeInOut(
                *fractions
        ))
        return this
    }

    fun duration(duration: Long): SpriteAnimatorBuilder {
        this.duration = duration
        return this
    }

    fun repeatCount(repeatCount: Int): SpriteAnimatorBuilder {
        this.repeatCount = repeatCount
        return this
    }

    fun build(): ObjectAnimator {
        val holders = arrayOfNulls<PropertyValuesHolder>(propertyValuesHolders.size)

        val proArray: Array<PropertyValuesHolder> = propertyValuesHolders.toTypedArray()

        val animator = ObjectAnimator.ofPropertyValuesHolder(sprite, *proArray)
        animator.duration = duration
        animator.repeatCount = repeatCount
        animator.interpolator = interpolator
        return animator
    }

}